include/type.hpp

```diff
@@ -168,4 +174,43 @@ class FuncType : public Type {
   bool ConvertibleHook_(const Type& that) const noexcept override;
 };
 
+class StructType : public Type {
+ public:
+  explicit StructType(std::vector<std::unique_ptr<Type>> field_types)
+      : field_types_{std::move(field_types)} {}
+
+  bool IsStruct() const noexcept override {
+    return true;
+  }
+
+  bool IsEqual(const Type& that) const noexcept override;
+  std::size_t size() const override;
+  std::string ToString() const override;
+  std::unique_ptr<Type> Clone() const override;
+
+ private:
+  std::vector<std::unique_ptr<Type>> field_types_;
+};
+
+class UnionType : public Type {
+ public:
+  /// @param selected_ The index of the selected field in the union.
+  explicit UnionType(std::vector<std::unique_ptr<Type>> field_types,
+                     std::size_t selected = 0)
```

Comment:
Concerning the design of the `UnionType`, it appears you're incorporating the selected index as part of the type information. However, this seems to present a semantic issue. Furthermore, in writing a C program, it is the programmer's responsibility to avoid accessing the field currently holding the value; the compiler is not obligated to manage this. In summary, I don't believe it's necessary to record which index is selected, particularly when dealing with the type itself.

Reply:
Make sense, I'll return the type size of its largest member.